lib/vnet/ssh_proxy.go

```diff
@@ -0,0 +1,260 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package vnet
+
+import (
+	"context"
+	"errors"
+	"log/slog"
+	"sync"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// sshConn represents an established SSH client or server connection.
+type sshConn struct {
+	conn  ssh.Conn
+	chans <-chan ssh.NewChannel
+	reqs  <-chan *ssh.Request
+}
+
+// proxySSHConnection transparently proxies SSH channels and requests
+// between 2 established SSH connections. serverConn represents an incoming SSH
+// connection where this proxy acts as a server, client represents an outgoing
+// SSH connection where this proxy acts as a client.
+func proxySSHConnection(
+	ctx context.Context,
+	serverConn sshConn,
+	clientConn sshConn,
+) {
+	closeConnections := func() {
+		clientConn.conn.Close()
+		serverConn.conn.Close()
+	}
+
+	// Avoid leaking goroutines by tracking them with a waitgroup.
+	var wg sync.WaitGroup
+	wg.Add(5)
+
+	// Close both connections if the context is canceled.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		defer wg.Done()
+		<-ctx.Done()
```

Comment:
Doesn't the `wg.Wait()` below block forever (or rather until `ctx` is externally cancelled, which might be never) since nothing else will call `cancel()` and the waitgroup is waiting on this `wg.Done()`?

Reply:
yeah my bad and my test conveniently missed this 🤦 I think context.AfterFunc solves it by avoiding a goroutine